Retain query plans

elonazoulay opened the pull request:
Modified SqlQueryManager and added QueryManagerConfig option to retain query plans.

ArturGajowy commented:
I see a property added, but no usage. Is this still a work in progress?

elonazoulay commented:
Hi @ArturGajowy - Thanks for catching this! I didn't add the SqlQueryManager.java change. Updating now.

ArturGajowy commented:
Reading from the code, the introduced property disables pruning expired queries whatsoever besides retaining the query plans. It might be that I think the property should be called `query.pruneExpired=off`. Looking deeper in the code it seems that pruning = removing query plan (and tasks, and stage), but we should state it consistently. I'd say either rename the property or the methods (from `pruneInfo` to, say `removePlan`).

martint commented:
What's the purpose of this change? If the goal is to retain just the plan, we should implement it in a way that does exactly that. This will cause the entire structure that holds the plan, stats, etc. to be kept around.

Also, we don't want an option to turn off pruning of expired queries. That would result in a memory leak. For that matter, I don't think this should be configurable, except if it's for one of these reasons:

1) during the transition period when changing semantics of existing functionality 
2) when new non-trivial functionality is introduced, as a way to turn it off in case there are problems (but only until the feature is proven)
3) as an escape hatch while the system is not smart enough to do the right thing (e.g., when it can't decide whether to apply certain optimizations)

elonazoulay commented:
Sounds good, thanks for looking at this! Will update shortly.

elonazoulay commented:
Updated to only retain the plan, and changed the option back to "query.retain-plan". 
We wanted to save the query plans so we can investigate performance regressions and correlate with a changed plan.
